tailfs/webdavfs/webdavfs.go

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package webdavfs provides an implementation of webdav.FileSystem backed by
+// a gowebdav.Client.
+package webdavfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/tailscale/gowebdav"
+	"github.com/tailscale/xnet/webdav"
+
+	"tailscale.com/tailfs/shared"
+	"tailscale.com/tstime"
+	"tailscale.com/types/logger"
+)
+
+const (
+	// keep requests from taking too long if the server is down or slow to respond
+	opTimeout = 2 * time.Second // TODO(oxtoacart): tune this
+)
+
+type Opts struct {
+	// Logf us a logging function to use for debug and error logging.
+	Logf logger.Logf
+	// URL is the base URL of the remote WebDAV server.
+	URL string
+	// Transport is the http.Transport to use for connecting to the WebDAV
+	// server.
+	Transport http.RoundTripper
+	// StatRoot, if true, will cause this filesystem to actually stat its own
+	// root via the remote server. If false, it will use a static directory
+	// info for the root to avoid a round-trip.
+	StatRoot bool
+	// StatCacheTTL, when greater than 0, enables caching of file metadata
+	StatCacheTTL time.Duration
+	// Clock, if specified, determines the current time. If not specified, we
+	// default to time.Now().
+	Clock tstime.Clock
+}
+
+// webdavFS adapts gowebdav.Client to webdav.FileSystem
+type webdavFS struct {
```

Comment:
Can we export this type? Obviously we need it to _implement_ the `webdav.FileSystem` interface, but the constructor can still return the concrete thing (that would make it easier to write integration tests too, though I realize we haven't got those yet).

Reply:
I think I see what you mean by an integration test, you mean actually testing something resembling tailscaled and the tailscale command, right?

Reply:
Right now our unit tests are all baked into the same package as the code. That's fine, and it's a pattern we've used all over the place, but if we want to build integration tests we'll need to import these types into a larger environment (as you say—maybe a whole node, or at least something where we have the ability to rub up against real calls from a client). Having the library types easy to stand up on their own, from another package, makes that a lot easier to do.